### PR TITLE
DEV-5039 | management command to fix payment method id (and details)

### PR DESCRIPTION
**Pull request by benjaminEwhite**
#### What's this PR do?

This PR adds a management command (`fix_contributions_missing_provider_payment_method_id`) that can be used to retrieve the relevant payment method id and details data for a given contribution (that has a either null or our dummy temp value for `provider_payment_method_id`.  When a relevant PM is found we update both provider_payment_method_id` and `provider_payment_method_details` on the relevant contribution.  

Additionally:

- Adds a new queryset method `ContributionQuerySet.exclude_disconnected_stripe_accounts` which excludes contributions with disconnected stripe accounts from a queryset.
- Removes `version` from our docker compose files, as we're now getting a warning about that property no longer being required.

#### Why are we doing this? How does it help us?

Fix a bunch of contributions in prod that. have bad payment method data, which blocks contributors from updating payment method and similar.

#### Are there detailed, specific, step-by-step testing instructions in the associated ticket?

Coming soon.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

No

#### What are the relevant tickets? Add a link to any relevant ones.

- [DEV-5039](https://news-revenue-hub.atlassian.net/browse/DEV-5039)
- [DEV-5328](https://news-revenue-hub.atlassian.net/browse/DEV-5328)
- [ITS-2600](https://news-revenue-hub.atlassian.net/browse/ITS-2600)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)? If so, open a ticket and link it to the ticket for this PR and list it here:

No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

-[ITS-2600](https://news-revenue-hub.atlassian.net/browse/ITS-2600)
- [DEV-5328](https://news-revenue-hub.atlassian.net/browse/DEV-5328)

[DEV-5039]: https://news-revenue-hub.atlassian.net/browse/DEV-5039?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+# This determines how many OR conditions we can have in a single stripe search query
+SEARCH_API_METADATA_ITEM_LIMIT = 10
+
+# otherwise we get spammed by stripe info logs when running this command
+logging.getLogger("stripe").setLevel(logging.ERROR)
+
+
+class Command(BaseCommand):
+    """Find contributions with missing or dummy val for provider_payment_method_id try to update.
```
Comment (njharman) [inline apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py:26]: "val" -> "value"

File: apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+                    ContributionStatus.CANCELED,
+                ]
+            )
+            .filter(
+                Q(provider_payment_method_id__isnull=True)
+                | Q(provider_payment_method_id=settings.DUMMY_PAYMENT_METHOD_ID)
+            )
+            .filter(
+                Q(interval=ContributionInterval.ONE_TIME, provider_payment_id__isnull=False)
+                | Q(~Q(interval=ContributionInterval.ONE_TIME), provider_subscription_id__isnull=False)
```
Comment (njharman) [inline apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py:61]: AFIK ~ is bitwise operator. Is it overloaded by Django Q()?  I did not figure out after few moments of searching.  If not, I don't believe we shouldn't be using bitwise here.

Does `interval != ContributionInter...` not work here?

**Review by njharman — CHANGES_REQUESTED** [review]

File: apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py (hunk 1 above)
Comment (benjaminEwhite) [reply apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py]: f182501902

File: apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py (hunk 2 above)
Comment (benjaminEwhite) [reply apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py:61]: This is idimoatic Django (and no you can't do `!=`).

https://docs.djangoproject.com/en/4.2/topics/db/queries/#complex-lookups-with-q-objects

File: apps/contributions/models.py [hunk 3]
```diff
@@ -233,6 +234,45 @@ def exclude_paymentless_canceled(self) -> models.QuerySet[Contribution]:
...
+            "Found %s eligible contribution%s to fix",
+            connected_contributions_count,
+            "" if connected_contributions_count == 1 else "s",
+        )
+        if ineligible_because_of_account.exists():
+            _inel_account = ineligible_because_of_account.count()
+            _plural = "" if _inel_account == 1 else "s"
+            logger.info(
+                "Found %s contribution%s with "
+                "null value for provider_payment_id that cannot be updated because account is disconnected or some other problem "
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/models.py]: Good catch. Copied this over from a more specific contex and missed cleaning up that log message.

File: apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py (hunk 2 above)
Comment (njharman) [reply apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py:61]: Thanks that's the docs I couldn't find / overloaded by Q().

File: apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py [hunk 4]
```diff
@@ -292,6 +292,26 @@ def process_contributions(
...
+        """
+        qs = Contribution.objects.filter(provider_payment_method_id=settings.DUMMY_PAYMENT_METHOD_ID).all()
+        if (
+            qs := Contribution.objects.filter(provider_payment_method_id=settings.DUMMY_PAYMENT_METHOD_ID).all()
+        ).exists():
+            self.stdout.write(
+                self.style.HTTP_INFO(f"Found {qs.count()} contributions with dummy payment method ID. Setting to null.")
+            )
+            with reversion.create_revision():
+                for con in Contribution.objects.filter(
```
Comment (njharman) [inline apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py:308]: Should we not (re)use the `qs` built above. Seems very WET, (un DRY).

Perhaps not convenient/possible here?; should get in habit of using batch versions of update/save.

File: apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py [hunk 5]
```diff
@@ -292,6 +292,26 @@ def process_contributions(
         )
         return updated_via_search_qs | updated_via_retrieve_qs, not_updated_via_search_qs | not_updated_via_retrieve_qs
 
+    def set_remaining_dummy_payment_method_ids_to_null(self) -> None:
+        """Set any remaining dummy payment method IDs to null.
+
+        If we need to reprocess, formerly dummy-valued-now-null-valued contributions will still be picked up.
+        """
+        qs = Contribution.objects.filter(provider_payment_method_id=settings.DUMMY_PAYMENT_METHOD_ID).all()
```
Comment (njharman) [inline apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py:300]: Duplicate.  `qs` is built twice here, and in if clause.

**Review by njharman — DISMISSED** [review]

File: apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py (hunk 4 above)
Comment (benjaminEwhite) [reply apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py:308]: I generally prefer batch update/save, but then you don't get reversion comments, and I wanted to ensure we have those.

File: apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py (hunk 5 above)
Comment (benjaminEwhite) [reply apps/contributions/management/commands/fix_contributions_missing_provider_payment_method_id.py]: bacfc09b4ce

**Review by x110dc — APPROVED** [review]
Approved.

**Pull request merged**